File: pkg/stream/udp_sender.go
package stream

import (
	"context"
	"fmt"
	"net"
	"os"
	"sync"
	"time"

	"github.com/sirupsen/logrus"
)

// UDPSender handles high-throughput UDP data transmission with 1024-byte payloads
type UDPSender struct {
	config     *StreamConfig
	conn       *net.UDPConn
	remoteAddr *net.UDPAddr
	stats      *StreamStats
	state      StreamState
	logger     *logrus.Logger
	
	// Link monitoring
	linkStatus     *LinkStatus
	linkMutex      sync.RWMutex
	lastHeartbeat  time.Time
	
	// Continuous mode
	continuousMode bool
	dataSink       string
	outputDir      string
	currentFile    *os.File
	fileMutex      sync.Mutex
	fileCounter    uint64
	
	// FEC components
	fecEncoder *FECEncoder
	
	// Flow control
	windowSize    int
	sentPackets   map[uint64]*Packet
	ackedPackets  map[uint64]bool
	windowMutex   sync.RWMutex
	
	// Performance optimization
	sendBuffer    chan *Packet
	ackBuffer     chan uint64
	
	// Control
	ctx           context.Context
	cancel        context.CancelFunc
	wg            sync.WaitGroup
}

// NewUDPSender creates a new UDP sender
func NewUDPSender(config *StreamConfig) *UDPSender {
	ctx, cancel := context.WithCancel(context.Background())
	
	// Initialize FEC if enabled
	var fecEncoder *FECEncoder
	if config.EnableFEC {
		fecConfig := FECConfig{
			Algorithm:  "xor",
			BlockSize:  100,
			Redundancy: config.FECRedundancy,
			MaxErrors:  10,
		}
		fecEncoder = NewFECEncoder(fecConfig, config.UDPPayloadSize)
	}
	
	return &UDPSender{
		config:         config,
		stats:          &StreamStats{StartTime: time.Now()},
		state:          StateIdle,
		logger:         logrus.New(),
		linkStatus:     &LinkStatus{IsConnected: false},
		continuousMode: config.ContinuousMode,
		dataSink:       config.DataSink,
		outputDir:      config.OutputDirectory,
		fecEncoder:     fecEncoder,
		windowSize:     config.WindowSize,
		sentPackets:    make(map[uint64]*Packet),
		ackedPackets:   make(map[uint64]bool),
		sendBuffer:     make(chan *Packet, config.BufferSize),
		ackBuffer:      make(chan uint64, config.BufferSize),
		ctx:            ctx,
		cancel:         cancel,
	}
}

// Connect establishes UDP connection to the receiver
func (s *UDPSender) Connect() error {
	s.state = StateConnecting
	
	// Resolve remote address
	remoteAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", s.config.RemoteAddr, s.config.Port))
	if err != nil {
		s.state = StateError
		return fmt.Errorf("failed to resolve remote address: %w", err)
	}
	s.remoteAddr = remoteAddr
	
	// Create UDP connection
	localAddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf("%s:%d", s.config.LocalAddr, 0))
	if err != nil {
		s.state = StateError
		return fmt.Errorf("failed to resolve local address: %w", err)
	}
	
	s.conn, err = net.DialUDP("udp", localAddr, remoteAddr)
	if err != nil {
		s.state = StateError
		return fmt.Errorf("failed to create UDP connection: %w", err)
	}
	
	// Set UDP options for high performance
	s.conn.SetWriteBuffer(s.config.BufferSize)
	s.conn.SetReadBuffer(s.config.BufferSize)
	
	s.state = StateConnected
	s.linkStatus.IsConnected = true
	s.lastHeartbeat = time.Now()
	s.logger.Info("UDP connection established")
	
	// Start background workers
	s.startBackgroundWorkers()
	
	return nil
}

// SendFile sends a file with 1024-byte UDP payloads
func (s *UDPSender) SendFile() error {
	if s.state != StateConnected {
		return fmt.Errorf("sender not connected")
	}
	
	s.state = StateTransferring
	s.logger.Info("Starting UDP file transfer")
	
	// Read and send file based on format
	switch s.config.FileFormat {
	case "bin":
		return s.sendBinaryFileUDP()
	case "pcap":
		return s.sendPcapFileUDP()
	default:
		return ErrInvalidFileFormat
	}
}

// sendBinaryFileUDP sends a binary file using 1024-byte UDP payloads
func (s *UDPSender) sendBinaryFileUDP() error {
	file, err := os.Open(s.config.InputFile)
	if err != nil {
		return fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()
	
	// Get file size for progress tracking
	fileInfo, err := file.Stat()
	if err != nil {
		return fmt.Errorf("failed to get file info: %w", err)
	}
	
	totalBytes := fileInfo.Size()
	s.logger.Infof("Sending binary file via UDP: %s (%d bytes)", s.config.InputFile, totalBytes)
	
	// Use fixed 1024-byte payload size for UDP
	payloadSize := s.config.UDPPayloadSize
	if payloadSize == 0 {
		payloadSize = DefaultUDPPayloadSize
	}
	
	// Read file in chunks and send
	sequenceNum := uint64(0)
	buffer := make([]byte, payloadSize)
	
	for {
		n, _ := file.Read(buffer)
		if n == 0 {
			break
		}
		
		// Create packet with actual data (pad to 1024 bytes if needed)
		payload := make([]byte, payloadSize)
		copy(payload, buffer[:n])
		
		flags := uint16(FlagUDPPayload)
		if sequenceNum == 0 {
			flags |= FlagCompressed
		}
		
		packet := CreatePacket(sequenceNum, payload, flags)
		
		// Send packet
		if err := s.sendUDPPacket(packet); err != nil {
			return fmt.Errorf("failed to send packet %d: %w", sequenceNum, err)
		}
		
		sequenceNum++
		s.stats.PacketsSent++
		s.stats.BytesSent += uint64(n)
		
		// Progress logging
		if sequenceNum%1000 == 0 {
			progress := float64(s.stats.BytesSent) / float64(totalBytes) * 100
			throughput := ConvertToGbps(CalculateThroughput(s.stats.BytesSent, time.Since(s.stats.StartTime)))
			s.logger.Infof("Progress: %.2f%%, Throughput: %.2f Gbps", progress, throughput)
		}
	}
	
	// Send end-of-stream packet
	endPacket := CreatePacket(sequenceNum, []byte{}, FlagEndOfStream|FlagUDPPayload)
	if err := s.sendUDPPacket(endPacket); err != nil {
		return fmt.Errorf("failed to send end packet: %w", err)
	}
	
	s.logger.Info("UDP file transfer completed")
	s.state = StateCompleted
	s.stats.EndTime = time.Now()
	
	return nil
}

// sendPcapFileUDP sends a pcap file using 1024-byte UDP payloads
func (s *UDPSender) sendPcapFileUDP() error {
	// Implementation for pcap file sending via UDP
	// Similar to binary file but with pcap-specific handling
	return s.sendBinaryFileUDP() // Simplified for now
}

// sendUDPPacket sends a single UDP packet with flow control
func (s *UDPSender) sendUDPPacket(packet *Packet) error {
	// Wait for window space
	for {
		s.windowMutex.RLock()
		windowUsed := len(s.sentPackets)
		s.windowMutex.RUnlock()
		
		if windowUsed < s.windowSize {
			break
		}
		
		// Wait a bit before checking again
		time.Sleep(100 * time.Microsecond)
	}
	
	// Serialize packet
	packetData := SerializePacket(packet)
	
	// Send packet via UDP
	_, err := s.conn.Write(packetData)
	if err != nil {
		return err
	}
	
	// Track sent packet for flow control
	s.windowMutex.Lock()
	s.sentPackets[packet.Header.SequenceNum] = packet
	s.windowMutex.Unlock()
	
	// Start retransmission timer
	go s.startRetransmissionTimer(packet.Header.SequenceNum)
	
	return nil
}

// startRetransmissionTimer starts a timer for packet retransmission
func (s *UDPSender) startRetransmissionTimer(sequenceNum uint64) {
	timer := time.NewTimer(s.config.RetryInterval)
	defer timer.Stop()
	
	select {
	case <-timer.C:
		// Check if packet was acknowledged
		s.windowMutex.RLock()
		acked := s.ackedPackets[sequenceNum]
		s.windowMutex.RUnlock()
		
		if !acked {
			// Retransmit packet
			s.windowMutex.RLock()
			packet := s.sentPackets[sequenceNum]
			s.windowMutex.RUnlock()
			
			if packet != nil {
				s.logger.Debugf("Retransmitting UDP packet %d", sequenceNum)
				packet.Header.Flags |= FlagRetransmit
				s.sendUDPPacket(packet)
				s.stats.PacketsRetransmitted++
			}
		}
	case <-s.ctx.Done():
		return
	}
}

// startBackgroundWorkers starts background goroutines
func (s *UDPSender) startBackgroundWorkers() {
	// ACK processor
	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		s.processUDPACKs()
	}()
	
	// Heartbeat sender
	s.wg.Add(1)
	go func() {
		defer s.wg.Done()
		s.sendHeartbeats()
	}()
	
	// Link monitor
	if s.config.EnableLinkMonitoring {
		s.wg.Add(1)
		go func() {
			defer s.wg.Done()
			s.monitorLink()
		}()
	}
}

// processUDPACKs processes incoming acknowledgments
func (s *UDPSender) processUDPACKs() {
	buffer := make([]byte, 1024)
	
	for {
		select {
		case <-s.ctx.Done():
			return
		default:
			// Set read timeout
			s.conn.SetReadDeadline(time.Now().Add(100 * time.Millisecond))
			
			n, _, err := s.conn.ReadFromUDP(buffer)
			if err != nil {
				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
					continue
				}
				s.logger.Errorf("Error reading UDP ACK: %v", err)
				return
			}
			
			// Process ACK data
			if n >= 8 {
				ackNum := uint64(0)
				for i := 0; i < 8; i++ {
					ackNum = (ackNum << 8) | uint64(buffer[i])
				}
				
				s.windowMutex.Lock()
				s.ackedPackets[ackNum] = true
				delete(s.sentPackets, ackNum)
				s.windowMutex.Unlock()
			}
		}
	}
}

// sendHeartbeats sends periodic heartbeat packets
func (s *UDPSender) sendHeartbeats() {
	ticker := time.NewTicker(s.config.HeartbeatInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			heartbeat := CreatePacket(0, []byte{}, FlagHeartbeat|FlagUDPPayload)
			s.sendUDPPacket(heartbeat)
			s.lastHeartbeat = time.Now()
		case <-s.ctx.Done():
			return
		}
	}
}

// monitorLink monitors link status and handles interruptions
func (s *UDPSender) monitorLink() {
	ticker := time.NewTicker(s.config.LinkMonitorInterval)
	defer ticker.Stop()
	
	for {
		select {
		case <-ticker.C:
			s.checkLinkStatus()
		case <-s.ctx.Done():
			return
		}
	}
}

// checkLinkStatus checks if the link is still active
func (s *UDPSender) checkLinkStatus() {
	s.linkMutex.Lock()
	defer s.linkMutex.Unlock()
	
	timeSinceHeartbeat := time.Since(s.lastHeartbeat)
	if timeSinceHeartbeat > s.config.LinkTimeout {
		if s.linkStatus.IsConnected {
			s.logger.Warn("Link interruption detected")
			s.state = StateLinkInterrupted
			s.linkStatus.IsConnected = false
			s.linkStatus.Interruptions++
			s.stats.LinkInterruptions++
		}
	} else {
		if !s.linkStatus.IsConnected {
			s.logger.Info("Link restored, attempting resync")
			s.attemptResync()
		}
	}
}

// attemptResync attempts to resync after link restoration
func (s *UDPSender) attemptResync() {
	s.state = StateResyncing
	
	// Send resync packet
	resyncPacket := CreatePacket(0, []byte{}, FlagResync|FlagUDPPayload)
	s.sendUDPPacket(resyncPacket)
	
	// Wait for acknowledgment
	time.Sleep(100 * time.Millisecond)
	
	s.linkStatus.IsConnected = true
	s.linkStatus.LastResync = time.Now()
	s.stats.ResyncCount++
	s.state = StateTransferring
	
	s.logger.Info("Resync completed successfully")
}

// GetStats returns current transfer statistics
func (s *UDPSender) GetStats() *StreamStats {
	s.stats.Throughput = CalculateThroughput(s.stats.BytesSent, time.Since(s.stats.StartTime))
	s.stats.LastLinkStatus = s.linkStatus.IsConnected
	return s.stats
}

// Close closes the UDP sender and cleans up resources
func (s *UDPSender) Close() error {
	s.cancel()
	s.wg.Wait()
	
	if s.conn != nil {
		return s.conn.Close()
	}
	return nil
}

File: src/udp_sender.cpp
#include "udp_sender.hpp"
#include "utils.cpp"
#include <iostream>
#include <fstream>
#include <cstring>
#include <algorithm>
#include <chrono>
#include <thread>

namespace hts {

UDPSender::UDPSender(const StreamConfig& config)
    : config_(config)
    , stats_()
    , state_(StreamState::IDLE)
    , socket_fd_(-1)
    , connected_(false)
    , window_size_(config.window_size)
    , running_(false)
    , sequence_num_(0) {
    
    stats_.start_time = std::chrono::steady_clock::now();
    link_status_.is_connected = false;
    link_status_.last_heartbeat = std::chrono::steady_clock::now();
}

UDPSender::~UDPSender() {
    close();
}

ErrorCode UDPSender::connect() {
    if (state_ != StreamState::IDLE) {
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    state_ = StreamState::CONNECTING;
    
    // Create UDP socket
    socket_fd_ = socket(AF_INET, SOCK_DGRAM, 0);
    if (socket_fd_ < 0) {
        log_error("Failed to create UDP socket");
        state_ = StreamState::ERROR;
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    // Set socket options for high performance
    int optval = 1;
    setsockopt(socket_fd_, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
    
    // Set buffer sizes
    setsockopt(socket_fd_, SOL_SOCKET, SO_SNDBUF, &config_.buffer_size, sizeof(config_.buffer_size));
    setsockopt(socket_fd_, SOL_SOCKET, SO_RCVBUF, &config_.buffer_size, sizeof(config_.buffer_size));
    
    // Set up remote address
    memset(&remote_addr_, 0, sizeof(remote_addr_));
    remote_addr_.sin_family = AF_INET;
    remote_addr_.sin_port = htons(config_.port);
    
    if (inet_pton(AF_INET, config_.remote_addr.c_str(), &remote_addr_.sin_addr) <= 0) {
        log_error("Invalid remote address: " + config_.remote_addr);
        close();
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    // Bind to local address
    struct sockaddr_in local_addr;
    memset(&local_addr, 0, sizeof(local_addr));
    local_addr.sin_family = AF_INET;
    local_addr.sin_addr.s_addr = htonl(INADDR_ANY);
    local_addr.sin_port = htons(0); // Let OS choose port
    
    if (bind(socket_fd_, (struct sockaddr*)&local_addr, sizeof(local_addr)) < 0) {
        log_error("Failed to bind UDP socket");
        close();
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    connected_ = true;
    state_ = StreamState::CONNECTED;
    link_status_.is_connected = true;
    last_heartbeat_ = std::chrono::steady_clock::now();
    
    log_info("UDP connection established to " + config_.remote_addr + ":" + std::to_string(config_.port));
    
    // Start background workers
    running_ = true;
    
    if (config_.enable_link_monitoring) {
        link_monitor_thread_ = std::thread(&UDPSender::link_monitor, this);
    }
    
    heartbeat_thread_ = std::thread(&UDPSender::heartbeat_sender, this);
    ack_thread_ = std::thread(&UDPSender::ack_processor, this);
    
    return ErrorCode::SUCCESS;
}

ErrorCode UDPSender::send_file() {
    if (state_ != StreamState::CONNECTED) {
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    state_ = StreamState::TRANSFERRING;
    log_info("Starting UDP file transfer");
    
    // Open input file
    input_file_.open(config_.input_file, std::ios::binary);
    if (!input_file_.is_open()) {
        log_error("Failed to open input file: " + config_.input_file);
        return ErrorCode::FILE_NOT_FOUND;
    }
    
    // Send file based on format
    ErrorCode result;
    if (config_.file_format == "bin") {
        result = send_binary_file();
    } else if (config_.file_format == "pcap") {
        result = send_pcap_file();
    } else {
        log_error("Invalid file format: " + config_.file_format);
        return ErrorCode::INVALID_FILE_FORMAT;
    }
    
    input_file_.close();
    
    if (result == ErrorCode::SUCCESS) {
        // Send end-of-stream packet
        Packet end_packet = create_packet(sequence_num_++, std::vector<uint8_t>(), FLAG_END_OF_STREAM | FLAG_UDP_PAYLOAD);
        send_packet(end_packet);
        
        log_info("UDP file transfer completed");
        state_ = StreamState::COMPLETED;
        stats_.end_time = std::chrono::steady_clock::now();
    }
    
    return result;
}

ErrorCode UDPSender::send_binary_file() {
    // Get file size for progress tracking
    input_file_.seekg(0, std::ios::end);
    uint64_t total_bytes = input_file_.tellg();
    input_file_.seekg(0, std::ios::beg);
    
    log_info("Sending binary file via UDP: " + config_.input_file + " (" + std::to_string(total_bytes) + " bytes)");
    
    // Use fixed 1024-byte payload size for UDP
    uint32_t payload_size = config_.udp_payload_size;
    if (payload_size == 0) {
        payload_size = DEFAULT_UDP_PAYLOAD_SIZE;
    }
    
    std::vector<uint8_t> buffer(payload_size);
    uint64_t bytes_sent = 0;
    
    while (input_file_.good() && running_) {
        // Read chunk from file
        input_file_.read(reinterpret_cast<char*>(buffer.data()), payload_size);
        std::streamsize bytes_read = input_file_.gcount();
        
        if (bytes_read == 0) {
            break;
        }
        
        // Create payload (pad to 1024 bytes if needed)
        std::vector<uint8_t> payload(buffer.begin(), buffer.begin() + bytes_read);
        if (payload.size() < payload_size) {
            payload.resize(payload_size, 0);
        }
        
        // Create packet
        uint16_t flags = FLAG_UDP_PAYLOAD;
        if (sequence_num_ == 0) {
            flags |= FLAG_COMPRESSED;
        }
        
        Packet packet = create_packet(sequence_num_++, payload, flags);
        
        // Send packet
        ErrorCode result = send_packet(packet);
        if (result != ErrorCode::SUCCESS) {
            log_error("Failed to send packet " + std::to_string(sequence_num_ - 1));
            return result;
        }
        
        bytes_sent += bytes_read;
        
        // Progress logging
        if (sequence_num_ % 1000 == 0) {
            double progress = static_cast<double>(bytes_sent) / total_bytes * 100.0;
            update_throughput();
            double throughput_gbps = convert_to_gbps(stats_.throughput);
            log_info("Progress: " + std::to_string(progress) + "%, Throughput: " + std::to_string(throughput_gbps) + " Gbps");
        }
    }
    
    return ErrorCode::SUCCESS;
}

ErrorCode UDPSender::send_pcap_file() {
    // Simplified pcap file sending - similar to binary for now
    return send_binary_file();
}

ErrorCode UDPSender::send_packet(const Packet& packet) {
    // Wait for window space
    if (!wait_for_window_space()) {
        return ErrorCode::BUFFER_FULL;
    }
    
    // Serialize packet
    std::vector<uint8_t> packet_data = serialize_packet(packet);
    
    // Send packet via UDP
    ssize_t sent = sendto(socket_fd_, packet_data.data(), packet_data.size(), 0,
                         (struct sockaddr*)&remote_addr_, sizeof(remote_addr_));
    
    if (sent < 0) {
        log_error("Failed to send UDP packet: " + std::string(strerror(errno)));
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    if (static_cast<size_t>(sent) != packet_data.size()) {
        log_error("Incomplete UDP packet sent");
        return ErrorCode::CONNECTION_CLOSED;
    }
    
    // Track sent packet for flow control
    {
        std::lock_guard<std::mutex> lock(window_mutex_);
        sent_packets_[packet.header.sequence_num] = packet;
    }
    
    // Start retransmission timer
    start_retransmission_timer(packet.header.sequence_num);
    
    return ErrorCode::SUCCESS;
}

void UDPSender::ack_processor() {
    std::vector<uint8_t> buffer(1024);
    
    while (running_) {
        // Set receive timeout
        struct timeval tv;
        tv.tv_sec = 0;
        tv.tv_usec = 100000; // 100ms
        setsockopt(socket_fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));
        
        socklen_t addr_len = sizeof(remote_addr_);
        ssize_t received = recvfrom(socket_fd_, buffer.data(), buffer.size(), 0,
                                   (struct sockaddr*)&remote_addr_, &addr_len);
        
        if (received < 0) {
            if (errno == EAGAIN || errno == EWOULDBLOCK) {
                // Timeout, continue
                continue;
            }
            log_error("Error reading UDP ACK: " + std::string(strerror(errno)));
            break;
        }
        
        // Process ACK data
        if (received >= 8) {
            uint64_t ack_num = 0;
            for (int i = 0; i < 8; ++i) {
                ack_num = (ack_num << 8) | static_cast<uint8_t>(buffer[i]);
            }
            
            process_ack(ack_num);
        }
    }
}

void UDPSender::heartbeat_sender() {
    while (running_) {
        std::this_thread::sleep_for(config_.heartbeat_interval);
        
        if (!running_) break;
        
        Packet heartbeat = create_packet(0, std::vector<uint8_t>(), FLAG_HEARTBEAT | FLAG_UDP_PAYLOAD);
        send_packet(heartbeat);
        
        {
            std::lock_guard<std::mutex> lock(link_mutex_);
            last_heartbeat_ = std::chrono::steady_clock::now();
        }
    }
}

void UDPSender::link_monitor() {
    while (running_) {
        std::this_thread::sleep_for(config_.link_monitor_interval);
        
        if (!running_) break;
        
        check_link_status();
    }
}

void UDPSender::check_link_status() {
    std::lock_guard<std::mutex> lock(link_mutex_);
    
    auto time_since_heartbeat = std::chrono::steady_clock::now() - last_heartbeat_;
    if (time_since_heartbeat > config_.link_timeout) {
        if (link_status_.is_connected) {
            log_info("Link interruption detected");
            state_ = StreamState::LINK_INTERRUPTED;
            link_status_.is_connected = false;
            link_status_.interruptions++;
            stats_.link_interruptions++;
        }
    } else {
        if (!link_status_.is_connected) {
            log_info("Link restored, attempting resync");
            attempt_resync();
        }
    }
}

ErrorCode UDPSender::attempt_resync() {
    state_ = StreamState::RESYNCING;
    
    // Send resync packet
    Packet resync_packet = create_packet(0, std::vector<uint8_t>(), FLAG_RESYNC | FLAG_UDP_PAYLOAD);
    ErrorCode result = send_packet(resync_packet);
    
    if (result == ErrorCode::SUCCESS) {
        // Wait for acknowledgment
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        
        link_status_.is_connected = true;
        link_status_.last_resync = std::chrono::steady_clock::now();
        stats_.resync_count++;
        state_ = StreamState::TRANSFERRING;
        
        log_info("Resync completed successfully");
    }
    
    return result;
}

bool UDPSender::wait_for_window_space() {
    while (running_) {
        std::lock_guard<std::mutex> lock(window_mutex_);
        if (sent_packets_.size() < window_size_) {
            return true;
        }
        
        // Wait a bit before checking again
        std::this_thread::sleep_for(std::chrono::microseconds(100));
    }
    return false;
}

void UDPSender::start_retransmission_timer(uint64_t sequence_num) {
    // Start a timer for packet retransmission
    std::thread([this, sequence_num]() {
        std::this_thread::sleep_for(config_.retry_interval);
        
        if (!running_) return;
        
        // Check if packet was acknowledged
        bool acked = false;
        {
            std::lock_guard<std::mutex> lock(window_mutex_);
            acked = acked_packets_.find(sequence_num) != acked_packets_.end();
        }
        
        if (!acked) {
            // Retransmit packet
            std::lock_guard<std::mutex> lock(window_mutex_);
            auto it = sent_packets_.find(sequence_num);
            if (it != sent_packets_.end()) {
                log_debug("Retransmitting UDP packet " + std::to_string(sequence_num));
                Packet& packet = it->second;
                packet.header.flags |= FLAG_RETRANSMIT;
                send_packet(packet);
                stats_.packets_retransmitted++;
            }
        }
    }).detach();
}

void UDPSender::process_ack(uint64_t sequence_num) {
    std::lock_guard<std::mutex> lock(window_mutex_);
    acked_packets_[sequence_num] = true;
    sent_packets_.erase(sequence_num);
}

StreamStats UDPSender::get_stats() const {
    std::lock_guard<std::mutex> lock(control_mutex_);
    StreamStats stats = stats_;
    stats.throughput = calculate_throughput(stats.bytes_sent, 
        std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::steady_clock::now() - stats.start_time));
    stats.last_link_status = link_status_.is_connected;
    return stats;
}

void UDPSender::update_stats(uint64_t bytes_sent, uint64_t packets_sent) {
    std::lock_guard<std::mutex> lock(control_mutex_);
    stats_.bytes_sent += bytes_sent;
    stats_.packets_sent += packets_sent;
}

void UDPSender::update_throughput() {
    auto now = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(now - stats_.start_time);
    stats_.throughput = calculate_throughput(stats_.bytes_sent, duration);
}

void UDPSender::close() {
    running_ = false;
    
    if (ack_thread_.joinable()) {
        ack_thread_.join();
    }
    if (heartbeat_thread_.joinable()) {
        heartbeat_thread_.join();
    }
    if (link_monitor_thread_.joinable()) {
        link_monitor_thread_.join();
    }
    
    if (socket_fd_ >= 0) {
        ::close(socket_fd_);
        socket_fd_ = -1;
    }
    
    connected_ = false;
    state_ = StreamState::IDLE;
}

bool UDPSender::is_connected() const {
    return connected_ && link_status_.is_connected;
}

void UDPSender::log_info(const std::string& message) const {
    std::cout << "[INFO] " << message << std::endl;
}

void UDPSender::log_error(const std::string& message) const {
    std::cerr << "[ERROR] " << message << std::endl;
}

void UDPSender::log_debug(const std::string& message) const {
    if (config_.log_level == "debug") {
        std::cout << "[DEBUG] " << message << std::endl;
    }
}

} // namespace hts